flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
+        after the path of a dataset is reported to the control node that a
+        successful write to that path is not persisted to the backing loop
+        device.
+        """
+        evaluate_function = None
+
+        def callback(when_description):
+            if evaluate_function:
+                evaluate_function(when_description)
+        actual_blockdevice_manager = BlockDeviceManager()
+        proxy_blockdevice_manager = CallbackProxy(
+            actual_blockdevice_manager, callback)
+
+        # Set up a deployer with a blockdevice_manager that has hooks so we can
+        # inject code around interactions with the blockdevices on the system.
+        deployer = create_blockdevicedeployer(
+            self, node_uuid=self.NODE_UUID,
+            block_device_manager=proxy_blockdevice_manager)
+
+        testing_mountroot = mountroot_for_test(self)
+
+        def has_file_named(blockdevice, filename):
+            """
+            Returns whether a given blockdevice has a file with the given name.
+            """
+            test_mountpoint = testing_mountroot.child(str(uuid4()))
+            test_mountpoint.makedirs()
+            actual_blockdevice_manager.mount(blockdevice, test_mountpoint)
+            result = test_mountpoint.child(filename).exists()
+            actual_blockdevice_manager.unmount(blockdevice)
+            test_mountpoint.remove()
+            return result
+
+        # This is information about the dataset that is intentionally only
+        # populated once the path of the dataset is known by the cluster and
+        # thus could be known by an API client.
+        dataset_info_from_cluster = {
+            # The path of the dataset.
+            'paths': set(),
+            # The blockdevice of the dataset. This is used for back door
+            # verification that writes are persisted to the blockdevice.
+            'dev': None,
+        }
+
+        def invariant(when_description):
+            """
+            An invariant that should always hold. Specifically, after the path
+            to a dataset is known by the cluster then writes to that path
+            should either succeed and be persisted to the underlying block
+            device, or the write should fail.
+
+            Returns False if none of the known paths of the dataset can be
+            written to.
+            """
+            write_success = []
+            for path in dataset_info_from_cluster['paths']:
+                device = dataset_info_from_cluster['dev']
+                filename = str(uuid4())
+                try:
+                    path.child(filename).touch()
+                except OSError:
+                    # If we failed to write, that is okay in many scenarios. At
+                    # least the failing write won't lead to the end user
+                    # thinking that they have an actually persisted data. Note
+                    # the failure, but don't immediately fail the test.
```

Comment:
An OSError just means that we failed to write. That's an okay outcome (and in fact desirable if it happens after a umount). We are only trying to prevent the case where a write is successful, but it is not persisted to the backend.